US_USELESS_SUPPRESSION_ON_* should report which annotation is useless

Using spotbugs-maven-plugin 4.9.3.0 , I get warnings like:

```
[ERROR] Medium: Suppressing annotation on the method ... is unnecessary [...] At ...:[lines 74-77] US_USELESS_SUPPRESSION_ON_METHOD
```

However, if there is more than one suppressed annotation, it's a guessing game which annotation was actually useless.

Can the error include the useless annotation(s) please?

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -12,6 +12,10 @@ Currently the versioning policy of this project follows [Semantic Versioning v2.
 - Do not report `US_USELESS_SUPPRESSION_ON_METHOD` on generated methods ([#3350](https://github.com/spotbugs/spotbugs/issues/3350))
 - Rewrite some member in `ResourceValueFrame.java` to Enum ([#2061](https://github.com/spotbugs/spotbugs/issues/2061))
 
+### Added
+
+- Added the unnecessary annotation to the `US_USELESS_SUPPRESSION_ON_*` messages ([#3395](https://github.com/spotbugs/spotbugs/issues/3395))
+
 ## 4.9.3 - 2025-03-14
 ### Added
 

--- a/spotbugs-tests/src/test/java/edu/umd/cs/findbugs/SuppressionMatcherTest.java
+++ b/spotbugs-tests/src/test/java/edu/umd/cs/findbugs/SuppressionMatcherTest.java
@@ -149,4 +149,51 @@ class SuppressionMatcherTest {
         verify(bugReporter).reportBug(bugCaptor.capture());
         assertThat("Bug type", bugCaptor.getValue().getBugPattern().getType(), startsWith("US_USELESS_SUPPRESSION"));
     }
+
+    @Test
+    void uselessSuppressionOnClassMessage() {
+        ClassWarningSuppressor suppressor = new ClassWarningSuppressor("XYZ", SuppressMatchType.DEFAULT, new ClassAnnotation("java.lang.String"));
+        BugInstance bugInstance = suppressor.buildUselessSuppressionBugInstance(new UselessSuppressionDetector());
+        assertThat("Message", bugInstance.getMessage(), is("US: Suppressing annotation XYZ on the class java.lang.String is unnecessary"));
+    }
+
+    @Test
+    void uselessSuppressionOnFieldMessage() {
+        FieldWarningSuppressor suppressor = new FieldWarningSuppressor("XYZ",
+                SuppressMatchType.DEFAULT, new ClassAnnotation("java.lang.String"),
+                new FieldAnnotation("java.lang.String", "test", "test", false));
+
+        BugInstance bugInstance = suppressor.buildUselessSuppressionBugInstance(new UselessSuppressionDetector());
+        assertThat("Message", bugInstance.getMessage(), is("US: Suppressing annotation XYZ on the field java.lang.String.test is unnecessary"));
+    }
+
+    @Test
+    void uselessSuppressionOnMethodMessage() {
+        MethodWarningSuppressor suppressor = new MethodWarningSuppressor("XYZ",
+                SuppressMatchType.DEFAULT, new ClassAnnotation("java.lang.String"),
+                new MethodAnnotation("java.lang.String", "test", "()Ljava/lang/Object;)", false),
+                false);
+
+        BugInstance bugInstance = suppressor.buildUselessSuppressionBugInstance(new UselessSuppressionDetector());
+        assertThat("Message", bugInstance.getMessage(), is("US: Suppressing annotation XYZ on the method String.test() is unnecessary"));
+    }
+
+    @Test
+    void uselessSuppressionOnMethodParameterMessage() {
+        ParameterWarningSuppressor suppressor = new ParameterWarningSuppressor("XYZ",
+                SuppressMatchType.DEFAULT, new ClassAnnotation("java.lang.String"),
+                new MethodAnnotation("java.lang.String", "bar", "()Ljava/lang/Object;)", false),
+                0);
+
+        BugInstance bugInstance = suppressor.buildUselessSuppressionBugInstance(new UselessSuppressionDetector());
+        assertThat("Message", bugInstance.getMessage(), is(
+                "US: Suppressing annotation XYZ on the parameter 1 of the method String.bar() is unnecessary"));
+    }
+
+    @Test
+    void uselessSuppressionOnPackageMessage() {
+        PackageWarningSuppressor suppressor = new PackageWarningSuppressor("XYZ", SuppressMatchType.DEFAULT, "java.lang");
+        BugInstance bugInstance = suppressor.buildUselessSuppressionBugInstance(new UselessSuppressionDetector());
+        assertThat("Message", bugInstance.getMessage(), is("US: Suppressing annotation XYZ on the package java.lang is unnecessary"));
+    }
 }

--- a/spotbugs/etc/messages.xml
+++ b/spotbugs/etc/messages.xml
@@ -9376,7 +9376,7 @@ Using floating-point variables should not be used as loop counters, as they are 
 
   <BugPattern type="US_USELESS_SUPPRESSION_ON_CLASS">
     <ShortDescription>Useless suppression on a class</ShortDescription>
-    <LongDescription>Suppressing annotation on the class {0} is unnecessary</LongDescription>
+    <LongDescription>Suppressing annotation {1} on the class {0} is unnecessary</LongDescription>
     <Details>
       <![CDATA[<p>
       Suppressing annotations <code>&amp;SuppressFBWarnings</code> should be removed from the source code as soon as they are no more needed.
@@ -9387,7 +9387,7 @@ Using floating-point variables should not be used as loop counters, as they are 
 
   <BugPattern type="US_USELESS_SUPPRESSION_ON_FIELD">
     <ShortDescription>Useless suppression on a field</ShortDescription>
-    <LongDescription>Suppressing annotation on the field {1} is unnecessary</LongDescription>
+    <LongDescription>Suppressing annotation {2} on the field {1} is unnecessary</LongDescription>
     <Details>
       <![CDATA[<p>
       Suppressing annotations <code>&amp;SuppressFBWarnings</code> should be removed from the source code as soon as they are no more needed.
@@ -9400,7 +9400,7 @@ Using floating-point variables should not be used as loop counters, as they are 
 
   <BugPattern type="US_USELESS_SUPPRESSION_ON_METHOD">
     <ShortDescription>Useless suppression on a method</ShortDescription>
-    <LongDescription>Suppressing annotation on the method {1} is unnecessary</LongDescription>
+    <LongDescription>Suppressing annotation {2} on the method {1} is unnecessary</LongDescription>
     <Details>
       <![CDATA[<p>
       Suppressing annotations <code>&amp;SuppressFBWarnings</code> should be removed from the source code as soon as they are no more needed.
@@ -9413,7 +9413,7 @@ Using floating-point variables should not be used as loop counters, as they are 
 
   <BugPattern type="US_USELESS_SUPPRESSION_ON_METHOD_PARAMETER">
     <ShortDescription>Useless suppression on a method parameter</ShortDescription>
-    <LongDescription>Suppressing annotation on the parameter {2} of the method {1} is unnecessary</LongDescription>
+    <LongDescription>Suppressing annotation {3} on the parameter {2} of the method {1} is unnecessary</LongDescription>
     <Details>
       <![CDATA[<p>
       Suppressing annotations <code>&amp;SuppressFBWarnings</code> should be removed from the source code as soon as they are no more needed.
@@ -9426,7 +9426,7 @@ Using floating-point variables should not be used as loop counters, as they are 
 
   <BugPattern type="US_USELESS_SUPPRESSION_ON_PACKAGE">
     <ShortDescription>Useless suppression on a package</ShortDescription>
-    <LongDescription>Suppressing annotation on the package {0} is unnecessary</LongDescription>
+    <LongDescription>Suppressing annotation {1} on the package {0} is unnecessary</LongDescription>
     <Details>
       <![CDATA[<p>
       Suppressing annotation <code>&amp;SuppressFBWarnings</code> should be removed from the source code as soon as they are no more needed.

--- a/spotbugs/src/main/java/edu/umd/cs/findbugs/ClassWarningSuppressor.java
+++ b/spotbugs/src/main/java/edu/umd/cs/findbugs/ClassWarningSuppressor.java
@@ -24,7 +24,8 @@ public class ClassWarningSuppressor extends WarningSuppressor {
     @Override
     public BugInstance buildUselessSuppressionBugInstance(UselessSuppressionDetector detector) {
         return new BugInstance(detector, BUG_TYPE, PRIORITY)
-                .addClass(clazz.getClassDescriptor());
+                .addClass(clazz.getClassDescriptor())
+                .addString(bugPattern);
     }
 
     @Override

--- a/spotbugs/src/main/java/edu/umd/cs/findbugs/FieldWarningSuppressor.java
+++ b/spotbugs/src/main/java/edu/umd/cs/findbugs/FieldWarningSuppressor.java
@@ -41,6 +41,7 @@ public class FieldWarningSuppressor extends ClassWarningSuppressor {
     public BugInstance buildUselessSuppressionBugInstance(UselessSuppressionDetector detector) {
         return new BugInstance(detector, BUG_TYPE, PRIORITY)
                 .addClass(clazz.getClassDescriptor())
-                .addField(field);
+                .addField(field)
+                .addString(bugPattern);
     }
 }

--- a/spotbugs/src/main/java/edu/umd/cs/findbugs/MethodWarningSuppressor.java
+++ b/spotbugs/src/main/java/edu/umd/cs/findbugs/MethodWarningSuppressor.java
@@ -44,7 +44,8 @@ public class MethodWarningSuppressor extends ClassWarningSuppressor {
     public BugInstance buildUselessSuppressionBugInstance(UselessSuppressionDetector detector) {
         return new BugInstance(detector, BUG_TYPE, PRIORITY)
                 .addClass(clazz.getClassDescriptor())
-                .addMethod(method);
+                .addMethod(method)
+                .addString(bugPattern);
     }
 
     @Override

--- a/spotbugs/src/main/java/edu/umd/cs/findbugs/PackageWarningSuppressor.java
+++ b/spotbugs/src/main/java/edu/umd/cs/findbugs/PackageWarningSuppressor.java
@@ -37,6 +37,8 @@ public class PackageWarningSuppressor extends WarningSuppressor {
 
     @Override
     public BugInstance buildUselessSuppressionBugInstance(UselessSuppressionDetector detector) {
-        return new BugInstance(detector, BUG_TYPE, PRIORITY).addClass(packageName);
+        return new BugInstance(detector, BUG_TYPE, PRIORITY)
+                .addClass(packageName)
+                .addString(bugPattern);
     }
 }

--- a/spotbugs/src/main/java/edu/umd/cs/findbugs/ParameterWarningSuppressor.java
+++ b/spotbugs/src/main/java/edu/umd/cs/findbugs/ParameterWarningSuppressor.java
@@ -43,6 +43,7 @@ public class ParameterWarningSuppressor extends ClassWarningSuppressor {
         return new BugInstance(detector, BUG_TYPE, PRIORITY)
                 .addClass(clazz.getClassDescriptor())
                 .addMethod(method)
-                .addParameterAnnotation(register, LocalVariableAnnotation.PARAMETER_ROLE);
+                .addParameterAnnotation(register, LocalVariableAnnotation.PARAMETER_ROLE)
+                .addString(bugPattern);
     }
 }

